quinn/examples/common/mod.rs

```diff
@@ -0,0 +1,99 @@
+//! Commonly used code in most examples.
+
+use futures::{Future, Stream};
+use quinn::{
+    Certificate, CertificateChain, ClientConfig, ClientConfigBuilder, Endpoint, EndpointDriver,
+    Incoming, PrivateKey, ServerConfig, ServerConfigBuilder, TransportConfig,
+};
+use std::error::Error;
+use std::net::ToSocketAddrs;
+use std::sync::Arc;
+use tokio::runtime::current_thread;
+
+/// Constructs a QUIC endpoint configured for use a client only.
+///
+/// ## Args
+///
+/// - server_certs: list of trusted certificates.
+#[allow(unused)]
+pub fn make_client_endpoint<A: ToSocketAddrs>(
+    bind_addr: A,
+    server_certs: &Vec<Vec<u8>>,
+) -> Result<(Endpoint, EndpointDriver), Box<Error>> {
+    let client_cfg = configure_client(server_certs)?;
+    let mut endpoint_builder = Endpoint::new();
+    endpoint_builder.default_client_config(client_cfg);
+    let (driver, endpoint, _incoming) = endpoint_builder.bind(bind_addr)?;
+    Ok((endpoint, driver))
+}
+
+/// Constructs a QUIC endpoint configured to listen for incoming connections on a certain address
+/// and port.
+///
+/// ## Returns
+///
+/// - UDP socket driver
+/// - a sream of incoming QUIC connections
+/// - server certificate serialized into DER format
+pub fn make_server_endpoint<A: ToSocketAddrs>(
+    bind_addr: A,
+) -> Result<(EndpointDriver, Incoming, Vec<u8>), Box<Error>> {
+    let (server_config, server_cert) = configure_server()?;
+    let mut endpoint_builder = Endpoint::new();
+    endpoint_builder.listen(server_config);
+    let (driver, _endpoint, incoming) = endpoint_builder.bind(bind_addr)?;
+    Ok((driver, incoming, server_cert))
+}
+
+/// Attempts to read data from the stream hence keeps it alive.
+/// If stream gets dropped, the other end of connection gets closed prematurely.
+pub fn keep_connection_alive(incoming: quinn::IncomingStreams) {
+    let task = incoming
+        .map_err(|e| {
+            println!("Incoming stream closed due to: {}", e);
+        })
+        .for_each(move |stream| {
+            let stream = match stream {
+                quinn::NewStream::Bi(_bi) => panic!("Unexpected bidirectional stream here"),
+                quinn::NewStream::Uni(uni) => uni,
+            };
+            quinn::read_to_end(stream, 4096)
+                .map_err(|e| panic!("read_to_end() failed: {}", e))
+                .and_then(move |(_stream, _data)| Ok(()))
+        });
+    current_thread::spawn(task);
+}
+
+/// Builds default quinn client config and trusts given certificates.
+///
+/// ## Args
+///
+/// - server_certs: a list of trusted certificates in DER format.
+fn configure_client(server_certs: &Vec<Vec<u8>>) -> Result<ClientConfig, Box<Error>> {
```

Comment:
```suggestion
fn configure_client(server_certs: &[&[u8]]) -> Result<ClientConfig, Box<Error>> {
```

Unless there's a specific reason this needs to know about the container type? Seems like we're only passing in statically sized sets of certs anyway.